core/src/main/java/de/javakaffee/web/msm/storage/RedisStorageClient.java

```diff
@@ -0,0 +1,292 @@
+import redis.clients.jedis.exceptions.JedisConnectionException;
+
+import de.javakaffee.web.msm.NamedThreadFactory;
+
+/**
+ * Storage client backed by a Jedis client instance.
+ */
+public class RedisStorageClient implements StorageClient {
+    protected static final Log _log = LogFactory.getLog(RedisStorageClient.class);
+
+    private final String _host;
+    private final int _port;
+    private final boolean _ssl;
+    private final JedisPool _pool = new JedisPool();
+    private final ExecutorService _executor = Executors.newFixedThreadPool(
+            Runtime.getRuntime().availableProcessors(), new NamedThreadFactory("msm-redis-client"));
+
+    /**
+     * Creates a <code>MemcachedStorageClient</code> instance which connects to the given Redis URL.
+     * 
+     * @param redisUrl redis URL 
+     */
+    public RedisStorageClient(String redisUrl) {
+        if (redisUrl == null)
+            throw new NullPointerException("Param \"redisUrl\" may not be null");
+        
+        if (_log.isDebugEnabled())
+            _log.debug(String.format("Creating RedisStorageClient with URL \"%s\"", redisUrl));
+
+        // Support a Redis URL in the form "redis://hostname:port" or "rediss://" (for SSL connections) like the client "Lettuce" does
+        if (!(redisUrl.startsWith("redis://") || redisUrl.startsWith("rediss://")))
+            throw new IllegalArgumentException("Redis URL must start with \"redis://\" or \"rediss://\"");
+        
+        _ssl = redisUrl.startsWith("rediss://");
+
+        String hostNamePort = redisUrl.substring(redisUrl.indexOf('/') + 2);
+        int pos = hostNamePort.indexOf(':');
+        if (pos != -1) {
+            _host = hostNamePort.substring(0, pos);
+            try {
+                _port = Integer.parseInt(hostNamePort.substring(pos + 1));
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Error parsing port number in Redis URL");
+            }
+        } else {
+            _host = hostNamePort;
+            _port = 6379;
+        }
+    }
+    
+    @Override
+    public Future<Boolean> add(final String key, final int exp, final byte[] o) {
+        
+        if (_log.isDebugEnabled())
+            _log.debug(String.format("Adding key to Redis (key=%s, exp=%s, o=%s)", key, exp, o.getClass().getName()));
+        
+        return _executor.submit(new RedisCommandCallable<Boolean>() {
+            private volatile boolean _setCompleted;
+            
+            @Override protected Boolean execute(BinaryJedis jedis) throws Exception {
+                byte[] kb = keyBytes(key);
+                if (_setCompleted || jedis.setnx(kb, o) == 1) {
+                    _setCompleted = true; // make sure to not call setnx() a second time if connection fails
+                    if (exp == 0)
+                        return true;
+                    else
+                        return jedis.expire(kb, convertExp(exp)) == 1;
```

Comment:
Nevermind, jedis.setnx couldn't have caused the exception because _setCompleted wouldn't be true then. Instead jedis.expire must have caused the first exception. Therefore an existing session should not be expired by this call.